docs: azurerm backend updates for 1.11 release (#36641)
hashicorp/terraform: website/docs/language/backend/azurerm.mdx

  - System Assigned Managed Identity attached to Azure compute instance (agent / runner)
- Service Principal / App Registration with Client Secret
- Service Principal / App REgistration with Client Certificate
- User Account with Azure CLI only (for local development cycle)

These types can be supplied via inputs or via a pre-authenticated Azure CLI. We cover them in more depth in the following sections.

### Data Plane URI

In [most cases](https://learn.microsoft.com/en-us/azure/storage/common/storage-account-overview#standard-endpoints) this can simply be inferred from the `storage_account_name` and `container_name`. But if you are using the ['Azure DNS zone endpoints' feature](https://learn.microsoft.com/en-us/azure/storage/common/storage-account-overview#azure-dns-zone-endpoints-preview) the backend will need to lookup the data plane URI from the management plane. This requires the setting the `lookup_blob_endpoint` configuration option to `true` and the `Reader` role assignment on the storage account.

## Azure Active Directory

This method requires a valid Azure Active Directory principal and a predictable storage account data plane URI.

> ### Required Inputs

Reviewer: ```suggestion
### Required Configuration Options
```

- `subscription_id` - The subscription ID of the storage account is required to query the management plane. This is only required if `lookup_blob_endpoint` is set to `true`. If using Azure CLI, this can be inferred from the CLI session. This can also be set via the `ARM_SUBSCRIPTION_ID` environment variable.
- `resource_group_name` - The resource group name of the storage account is required to query the management plane. This is only required if `lookup_blob_endpoint` is set to `true`.

### Storage Account Required Role Assignments

The recommended data plane role assignments required for this method are either one of:

- `Storage Blob Data Owner` on the storage account container (Recommended)
- `Storage Blob Data Contributor` on the storage account

The recommended management plane role assignments required for this method are:

- `Reader` on the storage account *(Only required if `lookup_blob_endpoint` is set to `true`)*

### Azure Active Directory with OpenID Connect / Workload identity federation

> #### Required Inputs

Reviewer: ```suggestion
#### Required Configuration Options
```

    oidc_azure_service_connection_id = "00000000-0000-0000-0000-000000000000"  # Can also be set via `ARM_OIDC_AZURE_SERVICE_CONNECTION_ID` environment variable.
    use_azuread_auth                 = true                                    # Can also be set via `ARM_USE_AZUREAD` environment variable.
    tenant_id                        = "00000000-0000-0000-0000-000000000000"  # Can also be set via `ARM_TENANT_ID` environment variable.
    client_id                        = "00000000-0000-0000-0000-000000000000"  # Can also be set via `ARM_CLIENT_ID` environment variable.
    storage_account_name             = "abcd1234"                              # Can be passed via `-backend-config=`"storage_account_name=<storage account name>"` in the `init` command.
    container_name                   = "tfstate"                               # Can be passed via `-backend-config=`"container_name=<container name>"` in the `init` command.
    key                              = "prod.terraform.tfstate"                # Can be passed via `-backend-config=`"key=<blob key name>"` in the `init` command.
  }
}
```

### Azure Active Directory with Compute Attached Managed Identity

> #### Required Inputs

Reviewer: ```suggestion
#### Required Configuration Options
```

The following additional inputs are always required for this sub-type:

- `use_msi` - Set to `true` to use the managed identity to authenticate to the storage account data plane. This can also be set via the `ARM_USE_MSI` environment variable.

> #### Optional Inputs

Reviewer: ```suggestion
#### Optional Configuration Options
```

    use_msi              = true                                    # Can also be set via `ARM_USE_MSI` environment variable.
    use_azuread_auth     = true                                    # Can also be set via `ARM_USE_AZUREAD` environment variable.
    tenant_id            = "00000000-0000-0000-0000-000000000000"  # Can also be set via `ARM_TENANT_ID` environment variable.
    client_id            = "00000000-0000-0000-0000-000000000000"  # Can also be set via `ARM_CLIENT_ID` environment variable. Not required for System Assigned Managed Identity.
    storage_account_name = "abcd1234"                              # Can be passed via `-backend-config=`"storage_account_name=<storage account name>"` in the `init` command.
    container_name       = "tfstate"                               # Can be passed via `-backend-config=`"container_name=<container name>"` in the `init` command.
    key                  = "prod.terraform.tfstate"                # Can be passed via `-backend-config=`"key=<blob key name>"` in the `init` command.
  }
}
```

### Azure Active Directory with Azure CLI

You must have a pre-authenticated Azure CLI session using any supported method.

> #### Required Inputs

Reviewer: ```suggestion
#### Required Configuration Options
```

    client_certificate_path     = "/path/to/bundle.pfx"                   # Can also be set via `ARM_CLIENT_CERTIFICATE_PATH` environment variable.
    client_certificate_password = "************************************"  # Can also be set via `ARM_CLIENT_CERTIFICATE_PASSWORD` environment variable.
    storage_account_name        = "abcd1234"                              # Can be passed via `-backend-config=`"storage_account_name=<storage account name>"` in the `init` command.
    container_name              = "tfstate"                               # Can be passed via `-backend-config=`"container_name=<container name>"` in the `init` command.
    key                         = "prod.terraform.tfstate"                # Can be passed via `-backend-config=`"key=<blob key name>"` in the `init` command.
  }
}
```

## Access Key

This method requires you find the Access Key for the storage account and supply it to the backend configuration.

The Access Key is then used to directly authenticate to the storage account data plane.

This method is retained for backwards compatibility, but we do not recommend it for new workloads.

> ### Required Inputs

Reviewer: ```suggestion
### Required Configuration Options
```

The following inputs are always required for this method:

- `access_key` - The Access Key of the storage account is required to authenticate to the storage account data plane. This can also be set via the `ARM_ACCESS_KEY` environment variable.

> - `storage_account_name` - The name of the storage account where the state file blob resides in.

Reviewer: ```suggestion
- `storage_account_name` - The name of the storage account to write the state file blob to.
```

> - `container_name` - The name of the storage account container where the state file blob resides in.

Reviewer: ```suggestion
- `container_name` - The name of the storage account container to write the state file blob to.
```

> - `key` - The name of the blob within the storage account container that the state file will be stored in.

Reviewer: ```suggestion
- `key` - The name of the blob within the storage account container to write the state file to.
```

### Example Configuration

~> **Warning!** This method requires you to manage and rotate a secret. Consider using OIDC as a more secure approach.

```hcl
terraform {
  backend "azurerm" {
    access_key           = "abcdefghijklmnopqrstuvwxyz0123456789..."  # Can also be set via `ARM_ACCESS_KEY` environment variable.
    storage_account_name = "abcd1234"                                 # Can be passed via `-backend-config=`"storage_account_name=<storage account name>"` in the `init` command.
    container_name       = "tfstate"                                  # Can be passed via `-backend-config=`"container_name=<container name>"` in the `init` command.
    key                  = "prod.terraform.tfstate"                   # Can be passed via `-backend-config=`"key=<blob key name>"` in the `init` command.
  }
}
```

## SAS Token

This method requires you generate a SAS Token for the storage account container or blob and supply it to the backend configuration.

The SAS Token is then used to directly authenticate to the storage account data plane.

This method is retained for backwards compatibility, but we do not recommend it for new workloads.

> ### Required Inputs

Reviewer: ```suggestion
### Required Configuration Options
```

The following inputs are always required for this method:

- `sas_token` - The SAS Token for the storage account container or blob is required to authenticate to the storage account data plane. This can also be set via the `ARM_SAS_TOKEN` environment variable.

> - `storage_account_name` - The name of the storage account where the state file blob resides in.

Reviewer: ```suggestion
- `storage_account_name` - The name of the storage account to write the state file blob to.
```

> - `container_name` - The name of the storage account container where the state file blob resides in.

Reviewer: ```suggestion
- `container_name` - The name of the storage account container to write the state file blob to.
```

> - `key` - The name of the blob within the storage account container that the state file will be stored in.

Reviewer: ```suggestion
- `key` - The name of the blob within the storage account container to write the state file to.
```

terraform {
  backend "azurerm" {
    sas_token            = "abcdefghijklmnopqrstuvwxyz0123456789..."  # Can also be set via `ARM_SAS_TOKEN` environment variable.
    storage_account_name = "abcd1234"                                 # Can be passed via `-backend-config=`"storage_account_name=<storage account name>"` in the `init` command.
    container_name       = "tfstate"                                  # Can be passed via `-backend-config=`"container_name=<container name>"` in the `init` command.
    key                  = "prod.terraform.tfstate"                   # Can be passed via `-backend-config=`"key=<blob key name>"` in the `init` command.
  }
}
```

## Access Key Lookup

This method requires a valid Azure Active Directory principal and is a fallback for when Azure Active Directory authentication cannot be used on the storage account data plane.

This method queries the management plane to get the storage account Access Key and then uses that Access Key to authenticate to the storage account data plane. As such it requires elevated permissions on the storage account.

This method is retained for backwards compatibility, but we do not recommend it for new workloads.

> ### Required Inputs

Reviewer: ```suggestion
### Required Configuration Options
```

The following inputs are always required for this method:

- `tenant_id` - The tenant ID of the Azure Active Directory principal is required to authenticate to the storage account management and data plane. If using Azure CLI, this can be inferred from the CLI session. This can also be set via the `ARM_TENANT_ID` environment variable.
- `subscription_id` - The subscription ID of the storage account is required to query the management plane. If using Azure CLI, this can be inferred from the CLI session. This can also be set via the `ARM_SUBSCRIPTION_ID` environment variable.
- `resource_group_name` - The resource group name of the storage account is required to query the management plane.

> - `storage_account_name` - The name of the storage account where the state file blob resides in.

Reviewer: ```suggestion
- `storage_account_name` - The name of the storage account to write the state file blob to.
```

Similar callout to the same wording above, if this is not accurate please feel free to discard this suggestion.

> - `storage_account_name` - The name of the storage account where the state file blob resides in.

Reviewer: ```suggestion
- `storage_account_name` - The name of the storage account to write the state file blob to.
```

> - `container_name` - The name of the storage account container where the state file blob resides in.

Reviewer: ```suggestion
- `container_name` - The name of the storage account container to write the state file blob to.
```

> - `container_name` - The name of the storage account container where the state file blob resides in.

Reviewer: ```suggestion
- `container_name` - The name of the storage account container to write the state file blob to.
```

> - `key` - The name of the blob within the storage account container that the state file will be stored in.

Reviewer: ```suggestion
- `key` - The name of the blob within the storage account container to write the state file to.
```

> - `key` - The name of the blob within the storage account container that the state file will be stored in.

Reviewer: ```suggestion
- `key` - The name of the blob within the storage account container to write the state file to.
```

> ### Optional Inputs

Reviewer: ```suggestion
### Optional Configuration Options
```

> These optional inputs apply when [looking up the data plane URI](#data-plane-uri) from the management plane. They are not required when the data plane URI can be inferred from the `storage_account_name` and `container_name`.

Reviewer: ```suggestion
These optional configuration options apply when [looking up the data plane URI](#data-plane-uri) from the management plane. They are not required when the data plane URI can be inferred from the `storage_account_name` and `container_name`.
```

- `lookup_blob_endpoint` - Set to `true` to lookup the storage account data plane URI from the management plane. This is required if you are using the 'Azure DNS zone endpoints' feature. Defaults to `false`. This value can also be sourced from the `ARM_USE_DNS_ZONE_ENDPOINT` environment variable.

### Storage Account Required Role Assignments

The recommended data plane role assignments required for this method are either one of:

- `Storage Blob Data Owner` on the storage account container (Recommended)
- `Storage Blob Data Contributor` on the storage account

The recommended management plane role assignments required for this method are:

- `Reader` on the storage account
- `Storage Account Key Operator Service Role` on the storage account

### Access Key Lookup with OpenID Connect / Workload identity federation

OpenID Connect / Workload identity federation is the recommended method for this scenario.

> #### Required Inputs

Reviewer: ```suggestion
#### Required Configuration Options
```

> The following additional inputs are always required for this sub-type:

Reviewer: ```suggestion
The following additional configuration options are always required for this sub-type:
```

    subscription_id                  = "00000000-0000-0000-0000-000000000000"  # Can also be set via `ARM_SUBSCRIPTION_ID` environment variable.
    client_id                        = "00000000-0000-0000-0000-000000000000"  # Can also be set via `ARM_CLIENT_ID` environment variable.
    resource_group_name              = "StorageAccount-ResourceGroup"          # Can be passed via `-backend-config=`"resource_group_name=<resource group name>"` in the `init` command.
    storage_account_name             = "abcd1234"                              # Can be passed via `-backend-config=`"storage_account_name=<storage account name>"` in the `init` command.
    container_name                   = "tfstate"                               # Can be passed via `-backend-config=`"container_name=<container name>"` in the `init` command.
    key                              = "prod.terraform.tfstate"                # Can be passed via `-backend-config=`"key=<blob key name>"` in the `init` command.
  }
}
```

### Access Key Lookup with Compute Attached Managed Identity

> #### Required Inputs

Reviewer: ```suggestion
#### Required Configuration Options
```

> The following additional inputs are always required for this sub-type:

Reviewer: ```suggestion
The following additional configuration options are always required for this sub-type:
```

- `use_msi` - Set to `true` to use the managed identity to authenticate to the storage account data plane. This can also be set via the `ARM_USE_MSI` environment variable.

> #### Optional Inputs

Reviewer: ```suggestion
#### Optional Configuration Options
```

> The following additional inputs are optional for this sub-type:

Reviewer: ```suggestion
The following additional configuration options are optional for this sub-type:
```

    tenant_id            = "00000000-0000-0000-0000-000000000000"  # Can also be set via `ARM_TENANT_ID` environment variable. Azure CLI will fallback to use the connected tenant ID if not supplied.
    subscription_id      = "00000000-0000-0000-0000-000000000000"  # Can also be set via `ARM_SUBSCRIPTION_ID` environment variable. Azure CLI will fallback to use the connected subscription ID if not supplied.
    resource_group_name  = "StorageAccount-ResourceGroup"          # Can be passed via `-backend-config=`"resource_group_name=<resource group name>"` in the `init` command.
    storage_account_name = "abcd1234"                              # Can be passed via `-backend-config=`"storage_account_name=<storage account name>"` in the `init` command.
    container_name       = "tfstate"                               # Can be passed via `-backend-config=`"container_name=<container name>"` in the `init` command.
    key                  = "prod.terraform.tfstate"                # Can be passed via `-backend-config=`"key=<blob key name>"` in the `init` command.
  }
}
```

### Access Key Lookup with Client Secret

> We retain this method for backwards compatibility only, do not use it for any new workloads.

Reviewer: ```suggestion
Terraform retains this method for backwards compatibility only, do not use it for any new workloads.
```

    subscription_id      = "00000000-0000-0000-0000-000000000000"  # Can also be set via `ARM_SUBSCRIPTION_ID` environment variable.
    client_id            = "00000000-0000-0000-0000-000000000000"  # Can also be set via `ARM_CLIENT_ID` environment variable.
    client_secret        = "************************************"  # Can also be set via `ARM_CLIENT_SECRET` environment variable.
    resource_group_name  = "StorageAccount-ResourceGroup"          # Can be passed via `-backend-config=`"resource_group_name=<resource group name>"` in the `init` command.
    storage_account_name = "abcd1234"                              # Can be passed via `-backend-config=`"storage_account_name=<storage account name>"` in the `init` command.
    container_name       = "tfstate"                               # Can be passed via `-backend-config=`"container_name=<container name>"` in the `init` command.
    key                  = "prod.terraform.tfstate"                # Can be passed via `-backend-config=`"key=<blob key name>"` in the `init` command.
  }
}
```

### Access Key Lookup with Client Certificate

> We retain this method for backwards compatibility only, do not use it for any new workloads.

Reviewer: ```suggestion
Terraform retains this method for backwards compatibility only, do not use it for any new workloads.
```

At a high level, the `azurerm` backend has 5 methods it can use to authenticate to the storage account data plane:

- [Direct Azure Active Directory](#direct-azure-active-directory) **(Recommended)**
- [Indirect Azure Active Directory with URI Lookup](#indirect-azure-active-directory-with-uri-lookup) **(Recommended for use with Azure DNS zone endpoints)**
- [Indirect Access Key Lookup](#indirect-access-key-lookup) *(Not recommended for new workloads)*
- [Direct SAS Token](#direct-sas-token) *(Not recommended for new workloads)*
- [Direct Access Key](#direct-access-key) *(Not recommended for new workloads)*

## Direct Azure Active Directory

This method requires a valid Azure Active Directory principal and a predictable storage account data plane URI.

There are 5 sub-types of authentication supported with this method.

- OpenID Connect / Workload identity federation **(Recommended)**
  - User Assigned Managed Identity with Federated Credentials **(Recommended)**
  - Service Princial / App Registration with Federated Credentials

> - Compute Attached User or System Assigned Managed Identity

Author: The reason I added this is to differentiate it from a User Assigned Managed Identity used for Workload identity federation. I'm trying to come at this from the perspective of a novice who may not know the Managed Identity option only works with an Azure Compute instance. I'll take a look and see if I can move that information elsewhere.

  - System Assigned Managed Identity attached to Azure compute instance (agent / runner)
- Service Principal / App Registration with Client Secret
- Service Principal / App REgistration with Client Certificate
- User Account with Azure CLI only (for local development cycle)

These sub-types can be supplied via inputs or via a pre-authenticated Azure CLI.

### Required Inputs

The following inputs are always required for this method:

- `use_azuread_auth` - Set to `true` to use Azure Active Directory authentication to the storage account data plane.
- `tenant_id` - The tenant ID of the Azure Active Directory principal is required to authenticate to the storage account data plane.
- `storage_account_name` - The name of the storage account is required to read / update the data plane.
- `container_name` - The name of the container within the storage account is required to read / update the data plane.
- `key` - The name of the blob key within the container is required to read / update the data plane.

### Storage Account Required Role Assignments

The minimum data plane role assignments required for this method are either one of:

> - `Storage Blob Data Owner` on the storage account container (Recommended)

Author: Unless anything has changed, historically the Contributor permission does not work. It has to be owner to work.

> - `Storage Blob Data Owner` on the storage account container (Recommended)

Author: The new SDK method may no longer need the Owner permission, are you able to test that? It would be awesome if only Contributor is required now.

> - `Storage Blob Data Owner` on the storage account container (Recommended)

Author: @magodo To be super clear, historically:

- Owner is required for permissions at the storage account **Container** scope
- Contributor is required at the **Storage Account** scope

So, just to confirm before I update the docs, you have tested with permissions only applied at the container level and no permissions on the storage account itself?

> - `Storage Blob Data Owner` on the storage account container (Recommended)

Reviewer 2: @jaredfholgate Thanks for the confirmation! My account has the "Storage Blob Data Contributor" assigned at the subscription level, which is probably inherited to the SA scope. So you are probably right. Whist I'm a bit confused about how come different role is needed at different scope? Especially for the [Storage Blob Data Contributor](https://learn.microsoft.com/en-us/azure/role-based-access-control/built-in-roles/storage#storage-blob-data-contributor) role, it has the following definition:

```
      "actions": [
        "Microsoft.Storage/storageAccounts/blobServices/containers/delete",
        "Microsoft.Storage/storageAccounts/blobServices/containers/read",
        "Microsoft.Storage/storageAccounts/blobServices/containers/write",
        "Microsoft.Storage/storageAccounts/blobServices/generateUserDelegationKey/action"
      ],
      "dataActions": [
        "Microsoft.Storage/storageAccounts/blobServices/containers/blobs/delete",
        "Microsoft.Storage/storageAccounts/blobServices/containers/blobs/read",
        "Microsoft.Storage/storageAccounts/blobServices/containers/blobs/write",
        "Microsoft.Storage/storageAccounts/blobServices/containers/blobs/move/action",
        "Microsoft.Storage/storageAccounts/blobServices/containers/blobs/add/action"
      ],
```

Which seems to cover all the permissions required already, or no?

> - `Storage Blob Data Owner` on the storage account container (Recommended)

Author: I think the best approach will be to leave it as it is for now. I'll do some testing once the patch release goes out and if it turns out the permissions can be reduced on the container scope option, I will submit a new PR to update the docs. Does that sound reasonable?

The *Access Key* method can be used directly, by specifying the access key, or in combination with an Azure Active Directory principal (user, service principal or managed identity). To use an Access Key directly you must generate one for your state file blob and specify it in the backend configuration. If using an Azure Active Directory principal and you supply the resource group name, Terraform will attempt to generate an access key for the storage account, using the authenticated Azure Active Directory principal.

The *SAS Token* method can only be used directly. You must generate a SAS Token for your state file blob and pass it to the backend config.

The `azurerm` backend supports the following authentication scenarios to connect to the storage account, based on the configuration variables provided:

> | Authentication Method | Storage Account Authentication Type | Triggering Configuration† |

Reviewer 2: I regard the `Authentication Method` here represents the AAD Auth Method, which applies to the following cases:

1. Storage uses AAD auth: AAD Auth for Storage scope needed
2. Storage has not specified a auth method, fallbacks to getting the access key: AAD Auth for ARM scope needed
3. (Only after https://github.com/hashicorp/terraform/pull/36623) Specify `lookup_blob_endpoint`: AAD Auth for ARM scope needed

In each case, the user require to provide:

- tenant id
- Auth related info: e.g. client id, client secret, `use_xxx` toggles

But the `subscription_id` and `resource_group_name` are only necessary for 2 and 3, but not 1. It is even not needed if the `AAD Auth` is not used, i.e. access key or sas token.

With above, this table looks a bit confusing. Instead, can we extend based on the above 3 (hopefully 4) auth methods:

- access key
- sas token
- AAD Auth
- Unspecified (will try to use AAD auth and make a GET call to get the access key)

Each of them has different implications of permissions and the minimal configuration needed.

Based on this, some additional factors affects the permissions and configurations:
- Different AAD Auth method used implies the needed options
- (Only after https://github.com/hashicorp/terraform/pull/36623) `lookup_blob_endpoint` is specified

Based on this, hopefully we can have a table with the following shape:

|Storage Auth Method|Required Attributes|Minimal RBAC Role|
|-|-|-|
|Access Key|`access_key=...`|NA|
|SAS Token|`sas_token=...`|NA|
|AAD Auth|`use_azuread_auth=true` and attributes in *AAD Auth* section|Storage Blob Data Contributor/Reader|
|Unspecified|`subscription_id=...`, `resource_group_name=...` and attributes in *AAD Auth* section|`storageAccounts/listKeys`, `storageAccounts/read`|

Then we can put down the different combinations of AAD Auth in its own section, which basically include:

- Client ID + Client Secret
- Client ID + Client Certificate
- Managed Identity
- Azure CLI
- OIDC
- Github OIDC
- (Since v1.12.0) ADO OIDC

Each of them has different attributes required.

Lastly, we can add some new sections talking about:

- The additional config and permission needed for `use_blob_endpoint`, i.e. it requires `storageAccounts/read`, and AAD Auth (ARM scope)
- The `Azure CLI` auth method can infer the `subscription_id` if not specified

I can tell the value of flattened version of the document, whilst from a practical and maintainable perspective, we shall make the structure a bit organized/complicated..

* `tenant_id` - (Optional) The Tenant ID of the principal. This can also be sourced from the `ARM_TENANT_ID` environment variable.

* `use_azuread_auth` - (Optional) Whether Azure Active Directory Authentication for storage account data plane authentication. This can also be sourced from the `ARM_USE_AZUREAD` environment variable.

* `subscription_id` - (Optional) The Subscription ID of the storage account required for management plane authentication. This can also be sourced from the `ARM_SUBSCRIPTION_ID` environment variable.

* `resource_group_name` - (Optional) The Name of the Resource Group in which the Storage Account exists required for management plane authentication.

* `access_key` - (Optional) The Access Key used to authenticate to the storage account data plane with the [Direct Access Key](#direct-access-key) authenticaton method. This can also be sourced from the `ARM_ACCESS_KEY` environment variable.

* `sas_token` - (Optional) The SAS Token used to authenticate to the storage account data plane with the [Direct SAS Token](#direct-sas-token) authentication method. This can also be sourced from the `ARM_SAS_TOKEN` environment variable.

> * `use_cli` - (Optional) Should the Azure CLI session be used for authentication to the storage account management and data plane? This value can also be sourced from the `ARM_USE_CLI` environment variable.

Reviewer: ```suggestion
* `use_cli` - (Optional) Set to `true` to use the Azure CLI session for authentication to the storage account management and data plane. This value can also be sourced from the `ARM_USE_CLI` environment variable.
```

> * `use_oidc` - (Optional) Should OpenID Connect / Workload identity federation authentication be used for authentication to the storage account management and data plane? This can also be sourced from the `ARM_USE_OIDC` environment variable.

Reviewer: ```suggestion
* `use_oidc` - (Optional) Set to `true` to use OpenID Connect / Workload identity federation authentication for authentication to the storage account management and data plane. This can also be sourced from the `ARM_USE_OIDC` environment variable.
```

* `client_id` - (Optional) The Client ID of the Azure Active Directory Principal required for some authentication sub-types. This can also be sourced from the `ARM_CLIENT_ID` environment variable.

> * `ado_pipeline_service_connection_id` - (Optional) The Azure DevOps Pipeline Service Connection ID required for Open ID Connect / Workload identity federation authentication with Azure DevOps. This can also be sourced from the `ARM_ADO_PIPELINE_SERVICE_CONNECTION_ID` or `ARM_OIDC_AZURE_SERVICE_CONNECTION_ID` Environment Variables. The provider will look for values in this order and use the first it finds configured.

Reviewer: ```suggestion
* `ado_pipeline_service_connection_id` - (Optional) The Azure DevOps Pipeline Service Connection ID required for Open ID Connect / Workload identity federation authentication with Azure DevOps. This can also be sourced from the `ARM_ADO_PIPELINE_SERVICE_CONNECTION_ID` or `ARM_OIDC_AZURE_SERVICE_CONNECTION_ID` environment variables. The provider will look for values in this order and use the first it finds configured.
```

> * `oidc_request_url` - (Optional) The URL for the Open ID Connect provider from which to request an ID token. This is only required for advanced scenarios or third party integrations. This can also be sourced from the `ARM_OIDC_REQUEST_URL`, `ACTIONS_ID_TOKEN_REQUEST_URL` or `SYSTEM_OIDCREQUESTURI` Environment Variables. The provider will look for values in this order and use the first it finds configured.

Reviewer: ```suggestion
* `oidc_request_url` - (Optional) The URL for the Open ID Connect provider from which to request an ID token. This is only required for advanced scenarios or third party integrations. This can also be sourced from the `ARM_OIDC_REQUEST_URL`, `ACTIONS_ID_TOKEN_REQUEST_URL` or `SYSTEM_OIDCREQUESTURI` environment variables. The provider will look for values in this order and use the first it finds configured.
```

> * `oidc_request_token` - (Optional) The bearer token for the request to the Open ID Connect provider. This is only required for advanced scenarios or third party integrations. This can also be sourced from the `ARM_OIDC_REQUEST_TOKEN`, `ACTIONS_ID_TOKEN_REQUEST_TOKEN` or `SYSTEM_ACCESSTOKEN` Environment Variables. The provider will look for values in this order and use the first it finds configured.

Reviewer: ```suggestion
* `oidc_request_token` - (Optional) The bearer token for the request to the Open ID Connect provider. This is only required for advanced scenarios or third party integrations. This can also be sourced from the `ARM_OIDC_REQUEST_TOKEN`, `ACTIONS_ID_TOKEN_REQUEST_TOKEN` or `SYSTEM_ACCESSTOKEN` environment variables. The provider will look for values in this order and use the first it finds configured.
```

* `oidc_token` - (Optional) The ID Token when authenticating using OpenID Connect. This is only required for advanced scenarios or third party integrations. This can also be sourced from the `ARM_OIDC_TOKEN` environment variable.

* `oidc_token_file_path` - (Optional) The path to a file containing an ID token when authenticating using OpenID Connect. This is only required for advanced scenarios or third party integrations. This can also be sourced from the `ARM_OIDC_TOKEN_FILE_PATH` environment variable.

> * `use_aks_workload_identity` (Optional) Should Azure AKS Workload Identity be used for Authentication? This is only required for advanced scenarios or third party integrations. This can also be sourced from the `ARM_USE_AKS_WORKLOAD_IDENTITY` environment variable.

Reviewer: ```suggestion
* `use_aks_workload_identity` (Optional) Set to `true` to use Azure AKS Workload Identity for authentication. This is only required for advanced scenarios or third party integrations. This can also be sourced from the `ARM_USE_AKS_WORKLOAD_IDENTITY` environment variable.
```

> * `use_msi` - (Optional) Should a Compute Attached Managed Service Identity be used for authentication to the storage account management and data planes? This can also be sourced from the `ARM_USE_MSI` environment variable.

Reviewer: ```suggestion
* `use_msi` - (Optional) Set to `true` to use a Compute Attached Managed Service Identity for authentication to the storage account management and data planes. This can also be sourced from the `ARM_USE_MSI` environment variable.
```
